### PR TITLE
Add tests for layernorm and add op_handler functions

### LINKED ISSUE #3438: Add Support for More nn.Modules Layers.
### Problem Description

I have tried SHAP on my Vision Transformer Model, and it said that it does not recognize the following layers:
- nn.GELU
- nn.Identity
-  nn.LayerNorm

Could you please add support for those layers, and maybe for many others. Thank you kindly for your consideration!

### Alternative Solutions

_No response_

### Additional Context

_No response_

### Feature request checklist

- [X] I have checked the issue tracker for duplicate issues.
- [X] I'd be interested in making a PR to implement this feature

### PR DESCRIPTION
**Pull request by JulioHC00**
## Overview

Partially closes #3438  by adding support for LayerNorm and testing it

Description of the changes proposed in this pull request:

Adds LayerNorm to the op_handler dictionary with nonlinear_1d as well as Identity as passthrough. I've kept the structure of the tests as similar as possible to previous ones for consistency.

## Checklist

- [ x] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [ x] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: shap/explainers/_deep/deep_pytorch.py [hunk 1]
```diff
@@ -400,6 +401,7 @@ def nonlinear_1d(module, grad_input, grad_output):
 op_handler["BatchNorm2d"] = linear_1d
 op_handler["BatchNorm3d"] = linear_1d
 
+op_handler["LayerNorm"] = nonlinear_1d
```
Comment (CloseChoice) [inline shap/explainers/_deep/deep_pytorch.py:404]: Hey, thanks for the update. Unfortunately this is wrong. Sorry, I had a thought about this and the problem here is that multiple inputs (given one output) vary. So for instance, let's have a look at the [tensorflow implementation for multiplicative attribution](https://github.com/shap/shap/blob/master/shap/explainers/_deep/deep_tf.py#L782) and deduct it from the definition of shapley values:

$\phi_{i} = \sum_{S \subseteq N \setminus \{i\}} \frac{|S|!(|N|-|S|-1)!}{|N|!} \left( f(S \cup \{i\}) - f(S) \right)$

Now let's assume we have just $x, y$ (and therefore $b_{x}, b_{y}$ as baselines). 
We are looking for a multiplication (or any function that takes two arguments):
$f(x, y) = x \cdot y$
Then this formula becomes:
$\phi_{x} = \frac{1}{2} (f(x, y) - f(b_{x}, y)) + \frac{1}{2} (f(x, b_{y}) - f(b_{x}, b_{y}))$
We would need to do something similar for the layer norm BUT this is computationally expensive, so we might be able to use the approach that the tensorflow implementation takes for [`softmax`](https://github.com/shap/shap/blob/master/shap/explainers/_deep/deep_tf.py#L488) (yes, pytorch impl is wrong here too!). I don't fully understand this yet, but doesn't seem too complicated.

Sorry, I just figured this out too late. Really appreciate your effort. Let me know how you want to proceed with this.

**Review by CloseChoice — CHANGES_REQUESTED** [review]

File: shap/explainers/_deep/deep_pytorch.py (hunk 1 above)
Comment (JulioHC00) [reply shap/explainers/_deep/deep_pytorch.py:404]: Ah! It was too easy to be true. So, if I understand this correctly, in the original DeepLIFT paper they have the rescale rule which actually only applies to a **single input**. That's what's implemented in the `nonlinear_1d` function? And that's why this isn't valid for `LayerNorm` which takes several inputs. So we need to use the full equation to derive the proper function. Is this right? If so I think I understand and can try to see how to translate the way it's done in tensorflow to pytorch.

P.S. Does this mean that the `softmax` implementation for pytorch also needs to be fixed?

Comment (CloseChoice) [reply shap/explainers/_deep/deep_pytorch.py:404]: @JulioHC00 yes, that's right. And also we would need to fix the implementation for pytorch's `softmax` attribution. I believe that once we understand how the tensorflow softmax is implemented, we can apply this to pytorch LayerNorm and softmax as well.

Comment (JulioHC00) [reply shap/explainers/_deep/deep_pytorch.py:404]: @CloseChoice I guess the only part I'm confused about is how does this fit with the op_handlers? Don't these handle the gradients in the backward propagation? How does this relate to implementing the shap values 
equation?

Comment (rjbruin) [reply shap/explainers/_deep/deep_pytorch.py:404]: @CloseChoice When I integrate LayerNorm into a testcase and use the `op_handler['LayerNorm'] = softmax`, the testcase passes. Do we need to create a testcase that *does* show how Softmax/LayerNorm are broken? What would be in such a testcase?
